### PR TITLE
KAFKA-12928: Add a check whether the Task's statestore is actually a directory

### PR DESCRIPTION
**Pull request by dongjinleekr**
The first commit shows how to reproduce the problem, and the second commit is the fix.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by dongjinleekr** [comment]
@ableegoldman Please have a look when you are free. :pray:

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 1]
```diff
@@ -126,7 +126,7 @@ public StateDirectory(final StreamsConfig config, final Time time, final boolean
                 throw new ProcessorStateException(
                     String.format("base state directory [%s] doesn't exist and couldn't be created", stateDirName));
             }
-            if (!stateDir.exists() && !stateDir.mkdir()) {
+            if ((stateDir.exists() && !stateDir.isDirectory()) || (!stateDir.exists() && !stateDir.mkdir())) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: Please split this up into a separate check for `if ((stateDir.exists() && !stateDir.isDirectory())` and then throw an accurate exception, eg `state directory could not be created as there is an existing file with the same name`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 2]
```diff
@@ -230,18 +230,25 @@ public UUID initializeProcessId() {
...
+                                taskParentDir.getPath(), taskDir.getPath()));
+                    }
+                    if (!taskDir.exists() && !taskDir.mkdir()) {
+                        throw new ProcessorStateException(
+                            String.format("task directory [%s] doesn't exist and couldn't be created", taskDir.getPath()));
+                    }
                 }
-                if (!taskDir.exists() && !taskDir.mkdir()) {
+            } else {
+                if (!taskDir.isDirectory()) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: Same here, this exception message does not apply to the case this is trying to catch

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 3]
```diff
@@ -312,7 +319,7 @@ private boolean taskDirIsEmpty(final File taskDir) {
      */
     File globalStateDir() {
         final File dir = new File(stateDir, "global");
-        if (hasPersistentStores && !dir.exists() && !dir.mkdir()) {
+        if (hasPersistentStores && ((dir.exists() && !dir.isDirectory()) || (!dir.exists() && !dir.mkdir()))) {
             throw new ProcessorStateException(
                 String.format("global state directory [%s] doesn't exist and couldn't be created", dir.getPath()));
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: ditto here, please add a separate check and exception

**Review by ableegoldman — COMMENTED** [review]
This seems like kind of an extreme edge case. Honestly I'm not sure if we really need to go out of our way to check for the case in which a user has for some reason already created a non-directory file with the same name. 

But I suppose it doesn't hurt to add the checks. If we do though we should make sure to be more explicit  and throw an appropriate informative exception.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java (hunk 1 above)
Comment (showuon) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: +1

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java (hunk 2 above)
Comment (showuon) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: +1

Comment (showuon) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: nit: we could use `else if` here.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java [hunk 4]
```diff
@@ -236,6 +236,30 @@ public void shouldThrowProcessorStateException() throws IOException {
...
+
+        // Replace application's stateDir to regular file
+        Utils.delete(appDir);
+        appDir.createNewFile();
+
+        assertThrows(ProcessorStateException.class, () -> directory.getOrCreateDirectoryForTask(taskId));
+    }
+
+    @Test
+    public void shouldThrowProcessorStateExceptionIfAppDirOccupied() throws IOException {
```
Comment (showuon) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java]: I think we're testing `testDir` Occupied here, not `AppDir`.

**Review by showuon — COMMENTED** [review]
@dongjinleekr , I agree with @ableegoldman that there should be a rare case, but nice to have. Thanks.

**Comment by dongjinleekr** [comment]
@ableegoldman @showuon Here it is, with rebasing onto the latest trunk.

**Comment by dongjinleekr** [comment]
Rebased onto the latest trunk cc/ @ableegoldman

**Review by showuon — APPROVED** [review]
LGTM! Thank you for the PR.

**Review by ableegoldman — APPROVED** [review]
LGTM

**Pull request merged**